src/configuration/prima.rs

```diff
@@ -10,17 +10,29 @@ pub struct PrimaConfiguration {
     from_addr: String,
     namespace: String,
     environment: Environment,
+    tags: Vec<String>,
 }
 
 impl PrimaConfiguration {
     pub fn new(to_addr: &str, from_addr: &str, namespace: &str, environment: Environment) -> Self {
+        let env_str = environment.to_string();
         Self {
             to_addr: to_addr.to_string(),
             from_addr: from_addr.to_string(),
             namespace: namespace.to_string(),
             environment,
+            tags: vec![format!("env:{}", env_str)],
         }
     }
+
+    pub fn with_tag<T: ToString>(self, key: &str, value: &T) -> Self {
+        self.do_with_tag(key, &value.to_string())
+    }
+
+    fn do_with_tag(mut self, key: &str, value: &str) -> Self {
```

Comment:
I think it'd be cleaner to just do this in `with_tag`.

Reply:
Doing it like this to save some compile time/code size in downstream deps - thin generic wrappers around concrete implementations is generally good from what I understand.

Reply:
I agree with @MaeIsBad that it would be cleaner to do this in with_tag. Could you elaborate a bit more or maybe add a link that explain this?

Reply:
I don't really have a reference, this is just a result of how monomorphization works. The compiler has to generate a copy of any generic fn for every different T it is called with. The shorter the generic fn is, the less the use of generics impacts compile time and (optimizations depending) code size (although arguments about code size are usually irrelevant since we don't work on micro controllers). For this reason it can be useful to separate code that relies on the generic parameters (in this case the call to to_string) from the rest of the fn body.

For fairly trivial functions, it is less of an issue, but it also doesn't hurt, and in this case it's basically free to do. It's not `pub` so doesn't impact users, and I don't think there's any lack of clarity about what is happening.

Discussions of concepts like "clean" are usually a matter of taste, where the back and forth can run long. If this is blocking for you I'll change it, if it isn't I'll leave it the way it is.

Reply:
Rather than "cleaner", I'd call it "unnecessary indirection".
I understand your point about monomorphization, but I don't think the trade-off is worth here. The gain would extemely small and most projects will likely have only one or two concrete implementations of `with_tag`.

Reply:
I agree with @Fs00, but it's not "blocking" for me. In the end, as you said, it's a matter of taste.
I personally tend to prefer terse code to super optimized code, given that we are doing web development it's perfectly fine.

Reply:
Sure, removed, but I think we're spending an awful lot of time [staring at bike sheds](https://en.wiktionary.org/wiki/bikeshedding) here - all of this is non user facing etc.

@matteosister just to be clear, the type of application being developed is irrelevant, since the primary thing impacted by monomorphization is compile time.